src/System.Management.Automation/engine/parser/PSType.cs

```diff
@@ -1091,11 +1091,16 @@ internal static List<DefineEnumHelper> Sort(List<DefineEnumHelper> defineEnumHel
 
             internal void DefineEnum()
             {
+                var underlyingType = _enumDefinitionAst.BaseTypes.Count() == 0 ? typeof(int) : _enumDefinitionAst.BaseTypes[0].TypeName.GetReflectionType();
+
                 var definedEnumerators = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
-                var enumBuilder = _moduleBuilder.DefineEnum(_typeName, Reflection.TypeAttributes.Public, typeof(int));
+                var enumBuilder = _moduleBuilder.DefineEnum(_typeName, Reflection.TypeAttributes.Public, underlyingType);
                 DefineCustomAttributes(enumBuilder, _enumDefinitionAst.Attributes, _parser, AttributeTargets.Enum);
-                int value = 0;
+
+                dynamic value = 0;
+                dynamic maxValue = underlyingType.GetField("MaxValue", BindingFlags.Public | BindingFlags.Static | BindingFlags.GetField).GetValue(null);
```

Comment:
Given that there is a finite number of acceptable types that are permitted here, wouldn't using a type-based switch here outperform reflection for this lookup?

Reply:
I don't think performance is a major concern here - it's retrieved _once_ per `enum` defined.

Reply:
Every Little Bit Counts™️  😄 

(But yeah, I guess that makes sense!)

Reply:
I'm sure the _engine_ should get rid of slow patterns even if they are supposedly rarely used.
We could use the pattern (type.GetTypeCode() and TypeCode.Int16):
https://github.com/dotnet/corefx/blob/master/src/System.Linq.Expressions/src/System/Dynamic/Utils/TypeUtils.cs#L43-L57